### PR TITLE
Add checkstyle check and fix script for go deps

### PR DESCRIPTION
**Pull request by bduffany**
This PR adds a check to enforce the following:
* `go mod tidy -e` would produce no diffs to `go.mod` or `go.sum`
* `go.mod` has exactly 2 `require()` blocks, one for direct imports and one for indirect imports. See https://github.com/golang/go/issues/45965
* Updating `deps.bzl` via Gazelle would produce no diffs

Provides a script to automatically fix most errors detected, or provide instructions on how to manually fix (`tools/fix_go_deps.sh`)

Note: the `-e` flag to `go mod tidy` lets us avoid the need to add the generated protos to the repo before running `go mod tidy`. This flag is not supported in Go 1.15 (the version currently used in GH actions), so this PR also adds a step to upgrade the Go version to v1.17.

Example failing run (before running `tools/fix_go_deps.sh` in this PR): https://github.com/buildbuddy-io/buildbuddy/runs/4960976363?check_suite_focus=true -- the diff is huge here because `go.mod` is particularly whacky these days, but the diff should be much smaller in the future.

---

**Version bump**: Patch

### CODE REVIEW

**Review by vadimberezniker — APPROVED** [review]
nice :)

File: .github/workflows/checkstyle.yaml [hunk 1]
```diff
@@ -17,6 +17,11 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v1
 
+      - name: Setup go
+        uses: actions/setup-go@v2
+        with:
+          go-version: "1.17.2"
```
Comment (tylerwilliams) [inline .github/workflows/checkstyle.yaml:23]: do you want to leave a comment in WORKSPACE that points to this file and mentions these versions should be kept in sync?

File: tools/fix_go_deps.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
```
Comment (tylerwilliams) [inline tools/fix_go_deps.sh:1]: can you add an option to ./buildfix.sh that does this so we continue to have one script that will "ready" your build for review?

File: .github/workflows/checkstyle.yaml (hunk 1 above)
Comment (bduffany) [reply .github/workflows/checkstyle.yaml]: Done

File: tools/fix_go_deps.sh (hunk 2 above)
Comment (bduffany) [reply tools/fix_go_deps.sh:1]: Done, added a `--fix-go-deps` flag for this. Also added an `--all` flag that is the same as `--gazelle --fix-go-deps`.

**Comment by bduffany** [comment]
FYI, verified that this works on Mac.

**Review by tempoz — COMMENTED** [review]
This looks great! Thanks so much for doing this; when `go.mod` got out-of-date, `gopls` would give up and my IDE functionality would stop working until I went and ran go mod tidy again, so this is super helpful.

File: go.mod [hunk 3]
```diff
@@ -5,26 +5,27 @@ go 1.17
 replace (
 	github.com/firecracker-microvm/firecracker-go-sdk => github.com/tylerwilliams/firecracker-go-sdk v0.22.1
 	github.com/go-redsync/redsync/v4 v4.4.1 => github.com/bduffany/redsync/v4 v4.4.1-minimal
+	github.com/lni/dragonboat/v3 => github.com/tylerwilliams/dragonboat/v3 v3.3.4-rc2
```
Comment (tylerwilliams) [inline go.mod:8]: thankyou!

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Pull request merged**